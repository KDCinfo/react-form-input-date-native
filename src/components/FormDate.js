/*
    React Component > Form > Input Field > Type: Date

        - An HTML5 native Date input field with non-native fallback support as outlined in the MDN documentation

    Expected data input:

        this.props.entryId                   // [int] >= 0 | Used for: IDs - field-date- field-inputYear- field-inputMonth- field-inputDay-
        this.props.showNativeDate            // [bool] true|false | Used for: Show native date input (t) or individual y/m/d selectors (f)
        this.props.handleChangeDate(newDate) // yyyymmdd | A callback function that should handle the (newDate); API, local / session, etc.
        this.props.fullDateYearLeft          // yyyy-mm-dd        | Format used to populate the native date input field
        this.props.yearSelect                // yyyy              | Non-native fallback support - Year field
        this.props.monthSelect               // 01 (0-padded)     | Non-native fallback support - Month field
        this.props.daySelect                 // 01 (0-padded)     | Non-native fallback support - Day field
        this.props.days                      // [<option>1</option><option>2</option>,...] | An array whose length is based on year & month.
        this.props.isDateRequired            // [bool] true|false |

        Resources:

        months                               // ['Jan','Feb',...] | Allows for parental 'day' field manipulation.

    Usage:

        <DateInput
            entryId={entryEdit.fid}
            showNativeDate={this.state.showNativeDate}
            handleChangeDate={this.handleChangeDate.bind(this)}
            date={entryEdit.date}
            days={populateDays(entryEdit.date)}
            daySelect={entryEdit.date.substr(6, 2)}
            monthSelect={entryEdit.date.substr(4, 2)}
            yearSelect={entryEdit.date.substr(0, 4)}
            fullDateYearLeft={entryEdit.date.substr(0, 4) + '-' + entryEdit.date.substr(4, 2) + '-' + entryEdit.date.substr(6, 2)}
            isDateRequired={this.state.fieldRequired['date']}
            isDateDisabled={!isEdit}
        />

    Compiler:

        - Keith D Commiskey (2017-06)
        - https://github.com/KDCinfo/react-form-input-date-native
*/

import React from "react"
import PropTypes from 'prop-types'
import { FormGroup, ControlLabel, FormControl } from 'react-bootstrap'

import { months } from '../util/form-functions'

class DateInput extends React.Component {
    constructor(props) {
        super(props)

        const thisDate = new Date(),
              years = Array.from(Array(thisDate.getFullYear()-2000+11), (e,i)=>i+(2000))

        this.state = {
            years: years,
            nativePickerDisplay: 'none',
            fallbackPickerDisplay: 'none',
            fallbackLabelDisplay: 'none',
            fallbackRequired: false
        }
    }
    componentWillMount() {
        // console.log('componentWillMount [this.state] and [this.props]')
        // console.log(this.state)
        // console.log(this.props)

        if (this.props.showNativeDate) {
            // Show the native date picker (keep fallback date entries hidden)
            this.setState({nativePickerDisplay: 'block'})

        } else {
            // Show the fallback date entries (keep native date picker hidden)
            this.setState({fallbackPickerDisplay: 'inline-block'})
            this.setState({fallbackLabelDisplay: 'inline-block'})
            this.setState({fallbackRequired: true})
        }
    }
    handleChangeYearSelect(e) {
        const newDate = e.target.value + this.props.monthSelect + this.props.daySelect

        this.props.handleChangeDate(newDate)
    }
    handleChangeMonthSelect(e) {
        const newMonth = ('0' + (e.target.selectedIndex + 1)).slice(-2),
              newDate = this.props.yearSelect + newMonth + this.props.daySelect

        this.props.handleChangeDate(newDate)
    }
    handleChangeDaySelect(e) {
        const newDay = ('0' + e.target.value).slice(-2),
              newDate = this.props.yearSelect + this.props.monthSelect + newDay

        this.props.handleChangeDate(newDate)
    }
    handleChangeDateNative(e) {
        // Native date input value is: 2017-12-31 // Storing without '-'
        this.props.handleChangeDate(e.target.value.replace(/-/gi, ''))
    }
    render() {
        // The native input 'date' value is always formatted [yyyy-mm-dd]
        // <input className="hide" name="date-old" onChange={this.props.handleChange} value={this.props.date} type="date" title="yyyy-mm-dd" placeholder="yyyy-mm-dd" pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}" />
        return (
            <div>
                <div style={{display: this.state.nativePickerDisplay}} className="nativeDatePicker">
                    <FormGroup controlId={'field-date-' + this.props.entryId}>
                        <ControlLabel srOnly={this.props.isDateDisabled} bsClass="field-label"><span>Date:</span></ControlLabel>
                        <FormControl
                            disabled={this.props.isDateDisabled}
                            required={this.props.isDateRequired}
                            type="date"
                            placeholder="Date"
                            name="inputDate"
                            onChange={this.handleChangeDateNative.bind(this)}
                            value={this.props.fullDateYearLeft}
                        />
                        <FormControl.Feedback />
                    </FormGroup>
                </div>
                <p style={{display: this.state.fallbackLabelDisplay}} className="fallbackLabel hide">Enter Date:</p>
                <div style={{display: this.state.fallbackPickerDisplay}} className="fallbackDatePicker">
                    <FormGroup>
                        <ControlLabel srOnly={this.props.isDateDisabled} htmlFor={"field-inputYear-" + this.props.entryId}><span>Year</span></ControlLabel>
                        <FormControl
                            disabled={this.props.isDateDisabled}
                            required={this.state.fallbackRequired && this.props.isDateRequired}
                            componentClass="select"
                            placeholder="Year"
                            id={"field-inputYear-" + this.props.entryId}
                            name="inputYear"
                            onChange={this.handleChangeYearSelect.bind(this)}
                            value={parseFloat(this.props.yearSelect)}
                        >
                            { this.state.years.map( (day, idx) => <option key={idx}>{day}</option> ) }
                        </FormControl>
                    </FormGroup>
                    <FormGroup>
                        <ControlLabel srOnly={this.props.isDateDisabled} htmlFor={"field-inputMonth-" + this.props.entryId}><span>Month</span></ControlLabel>
                        <FormControl
                            disabled={this.props.isDateDisabled}
                            required={this.state.fallbackRequired && this.props.isDateRequired}
                            componentClass="select"
                            placeholder="Year"
                            id={"field-inputMonth-" + this.props.entryId}
                            name="inputMonth"
                            onChange={this.handleChangeMonthSelect.bind(this)}
                            value={months[parseFloat(this.props.monthSelect)-1]}
                        >
                            { months.map( (month, idx) => <option key={idx}>{month}</option> ) }
                        </FormControl>
                    </FormGroup>
                    <FormGroup>
                        <ControlLabel srOnly={this.props.isDateDisabled} htmlFor={"field-inputDay-" + this.props.entryId}><span>Day</span></ControlLabel>
                        <FormControl
                            disabled={this.props.isDateDisabled}
                            required={this.state.fallbackRequired && this.props.isDateRequired}
                            componentClass="select"
                            placeholder="Year"
                            id={"field-inputDay-" + this.props.entryId}
                            name="inputDay"
                            onChange={this.handleChangeDaySelect.bind(this)}
                            value={parseFloat(this.props.daySelect)}
                        >
                            { this.props.days.map( (day, idx) => <option key={idx}>{day}</option> ) }
                        </FormControl>
                    </FormGroup>
                </div>
            </div>
        )
    }
}

if(process.env.NODE_ENV !== 'production') {
    DateInput.propTypes = {

        entryId : PropTypes.number.isRequired,
        // [int] >= 0 | Used for: IDs - field-date- field-inputYear- field-inputMonth- field-inputDay-

        handleChangeDate : PropTypes.func,
        // yyyymmdd | A callback function that should handle the (newDate); API, local / session, etc.

        showNativeDate : PropTypes.bool.isRequired,
        // [bool] true|false | Used for: Show native date input (t) or individual y/m/d selectors (f)

        fullDateYearLeft : PropTypes.string.isRequired,
        // yyyy-mm-dd        | Format used to populate the native date input field

        yearSelect : PropTypes.string.isRequired,
        // yyyy              | Non-native fallback support - Year field

        monthSelect : PropTypes.string.isRequired,
        // 01 (0-padded)     | Non-native fallback support - Month field

        daySelect : PropTypes.string.isRequired,
        // 01 (0-padded)     | Non-native fallback support - Day field

        days : PropTypes.array.isRequired,
        // [<option>1</option><option>2</option>,...] | An array whose length is based on year & month.

        isDateRequired : PropTypes.bool.isRequired,
        // [bool] true|false |

        isDateDisabled : PropTypes.bool.isRequired
        // [bool] true|false |
    }
}
/*
    // PropTypes.string.isRequired
    // PropTypes.array
    // PropTypes.object
    // PropTypes.number
    // PropTypes.func
    // PropTypes.bool

    this.props.entryId                   // [int] >= 0 | Used for: IDs - field-date- field-inputYear- field-inputMonth- field-inputDay-
    this.props.showNativeDate            // [bool] true|false | Used for: Show native date input (t) or individual y/m/d selectors (f)
    this.props.handleChangeDate(newDate) // yyyymmdd | A callback function that should handle the (newDate); API, local / session, etc.
    this.props.fullDateYearLeft          // yyyy-mm-dd        | Format used to populate the native date input field
    this.props.yearSelect                // yyyy              | Non-native fallback support - Year field
    this.props.monthSelect               // 01 (0-padded)     | Non-native fallback support - Month field
    this.props.daySelect                 // 01 (0-padded)     | Non-native fallback support - Day field
    this.props.days                      // [<option>1</option><option>2</option>,...] | An array whose length is based on year & month.
    this.props.isDateRequired            // [bool] true|false |
    this.props.isDateDisabled
*/

export default DateInput