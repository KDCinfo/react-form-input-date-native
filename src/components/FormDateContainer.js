/*
    React Component > Form > Container

        - Parent container for: [Form > Input > Type: Date]
        - (native Date input field with non-native fallback)

    Simplified Overview:

        Native:
            <input type="date" />

        Non-Native Fallback:
            <select name="year" />
            <select name="month" />
            <select name="day" />

    Project Dependencies:

        {   "dependencies": {
                "prop-types": "^15.5.10",
                "react": "^15.6.1",
                "react-bootstrap": "^0.31.0",
                "react-dom": "^15.6.1"
            }
        }

    Integration / Implementation:

        - Pass in to this Component via a Prop:
            `formParams` = { fid: 0, date: '20170701' }

        - For updating the form's state (onSubmit):
            Search this file for: CUSTOM

    Compiler:

        - Keith D Commiskey (2017-06)

        This code is hosted and run on GitHub via Travis CI.

        - GitHub Source (Code) -- https://github.com/KDCinfo/react-form-input-date-native
        - GitHub Pages (Demo) -- https://KDCinfo.github.io/react-form-input-date-native/
        - Travis CI (Prod Build) -- https://travis-ci.org/KDCinfo/react-form-input-date-native
*/

import React from 'react'
import PropTypes from 'prop-types'
import { Form, Row, Col, Button, Panel, Well } from 'react-bootstrap'

import DateInput from './FormDate'

import { populateDays } from '../util/form-functions'

class FormDateContainer extends React.Component {
    constructor(props) {
        super(props)

        // Test whether a native date input falls back to a text input (i.e., native date input is not supported)
        //
            let testType = document.createElement('input'),
                showNativeDate
            testType.type = 'date'
            showNativeDate = testType.type === 'text' ? false : true

        this.state = {
            showNativeDate: showNativeDate,
            formState: this.initialFormData(),   // Updated on input field changes
            sourceState: this.initialFormData(), // Is only updated on form submit
            fieldRequired: {
                date: true
            },
            idValid: null,
            idError: '',
            dateValid: null,
            dateError: ''
        }

        // Bind component methods
        //
            this.validateForm = this.validateForm.bind(this)
            this.clearForm = this.clearForm.bind(this)
            this.resetState = this.resetState.bind(this)
    }
    initialFormData() {
        const newDate = new Date()
        return {
            "fid": 0,
            "date": (newDate.getFullYear()) + ('0' + (newDate.getMonth()+1)).slice(-2) + ('0' + newDate.getDate()).slice(-2),
        }
    }
    componentDidMount() {
        // Overwrite `sourceState` (initialFormData)
        // If object is passed in as a prop

        this.resetState()
    }
    resetState() {
        const testProps = this.props.formParams,        // 'formParams' Expects Object with a 'date' key
                                                        // { fid: [0], date: 'yyyymmdd' }
              propsHasDate = typeof(testProps) !== 'undefined' && typeof(testProps.date) !== 'undefined',
              thisProps = propsHasDate ? testProps : this.state.sourceState,
              mergedFormData = Object.assign({}, {...this.state.sourceState}, {...thisProps})

        this.setState({
            'sourceState': mergedFormData,
            'formState': mergedFormData
        })
    }
    clearForm() {
        // Only reset 'formState';
        // 'sourceState' is only changed (updated) on form submit

        this.setState({ formState: this.state.sourceState })
    }
    handleChange = (field, data) => {
        const oldFormData = this.state.formState,
              newFormData = {...oldFormData, [field]: data}
        this.setState({ 'formState': newFormData })
    }
    handleChangeDate(newDate) {
        this.handleChange('date', newDate)
    }
    handleSubmit(e) {
        if (e) {
            e.preventDefault()
        }

        const thisEntry = this.state.formState

        // Validate Form
        if (this.validateForm( thisEntry )) {

            // CUSTOM:

            // This is where you would pass this form's data [this.state.formData] to a prop's callback function
                // e.g., this.props.processFormData(this.state.formState)
                // Or call a Redux action to update the core Redux state
            // Instead, for this standalone purpose, we'll just update our fake 'sourceState'
            this.setState( { sourceState: thisEntry } )

            console.log('[UPDATE submitted]', thisEntry)

        } else {
            console.log('Errors were found -> Error states have been set (which should toggle individual field error visibility)')
        }
    }
    validateForm(form) {

        const errors = {},
              formatDateStripped = /^\d{8}$/

        for (let [key, val] of Object.entries( form )) {
            switch (key) {
                case 'fid':
                    // console.log('fid', val, typeof(val))

                    if (typeof(val) !== 'number' || (typeof(val) === 'number' && val < 0)) {
                        errors[key] = '[' + key + '] should be a Whole Number'
                        this.setState({ idValid: 'error' })
                        this.setState({ idError: errors[key] })
                    }
                    break

                case 'date':
                    // console.log('date', val, typeof(val))

                    if (typeof(val) !== 'string' || val.length !== 8 || !val.match(formatDateStripped)) {
                        errors[key] = 'The [' + key + '] field should be an 8-digit number representing yyyymmdd'
                        this.setState({ dateValid: 'error' })
                        this.setState({ dateError: errors[key] })
                    } else {
                        // This could be validated further where:
                            // yyyy should be a number between 1900 and 9999
                            // mm   should be a 2-digit string (number) and should represent a float from 1-12
                            // dd   should be a 2-digit string (number) and should represent a float from 1-31
                    }
                    break

                default:
                    break
            }
        }

        if (Object.keys(errors).length > 0) {
            return false
        } else {
            return true
        }
    }
    render() {

        const // { id, date } = this.props
              // { id, date } = this.state.sourceState,
              formId = this.state.formState.fid,
              propsObj = {
                'data-id': formId,
                required: true
              },
        isEdit = true,
        entryEdit = this.state.formState

        return (
            <Form onSubmit={this.handleSubmit.bind(this)} data-id={'form-' + entryEdit.fid} inline>
                <Row className="show-grid">
                    <Col className="field-col" xs={12}>
                        <DateInput
                            {...propsObj}
                            entryId={entryEdit.fid}
                            showNativeDate={this.state.showNativeDate}
                            handleChangeDate={this.handleChangeDate.bind(this)}
                            date={entryEdit.date}
                            days={populateDays(entryEdit.date)}
                            daySelect={entryEdit.date.substr(6, 2)}
                            monthSelect={entryEdit.date.substr(4, 2)}
                            yearSelect={entryEdit.date.substr(0, 4)}
                            fullDateYearLeft={entryEdit.date.substr(0, 4) + '-' + entryEdit.date.substr(4, 2) + '-' + entryEdit.date.substr(6, 2)}
                            isDateRequired={this.state.fieldRequired['date']}
                            isDateDisabled={!isEdit}
                        />
                        <div className="entry-buttons">
                            <Button className="btn btn-primary" type="submit">Submit Form</Button>
                            <Button className="btn btn-default" onClick={this.clearForm}>Reset Form</Button>
                        </div>
                    </Col>
                </Row>
                <Row>
                    <Col xs={12}>
                        <br/>
                        <Panel bsSize="xs"><Well>
                            <p>
                                This React Form Component demo is hosted and run&nbsp;
                                <a href="https://github.com/KDCinfo/react-form-input-date-native" target="kdcNewWin">on GitHub</a> via&nbsp;
                                <a href="https://travis-ci.org/KDCinfo/react-form-input-date-native" target="kdcNewWin">Travis CI</a>.
                            </p>
                            <ul>
                                <li><a href="https://github.com/KDCinfo/react-form-input-date-native" target="kdcNewWin"><b>https://github.com/KDCinfo/</b>react-form-input-date-native</a></li>
                                <li><a href="https://travis-ci.org/KDCinfo/react-form-input-date-native" target="kdcNewWin"><b>https://travis-ci.org/KDCinfo/</b>react-form-input-date-native</a></li>
                            </ul>
                        </Well></Panel>
                    </Col>
                </Row>
            </Form>
        )
    }
}

if(process.env.NODE_ENV !== 'production') {
    FormDateContainer.propTypes = {

        formParams : PropTypes.object.isRequired
        // Object with a 'date' key: { fid: 0, date: 'yyyymmdd' }

    }
}

export default FormDateContainer