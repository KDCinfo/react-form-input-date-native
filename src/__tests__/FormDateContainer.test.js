import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import FormDateContainer from '../components/FormDateContainer'
import FormDate from '../components/FormDate'

// The tests included below, although they do pass, are incomplete
// and meant only to serve as a base for testing (PRs are welcome)

  // Tests Include:
  // Jest Snapshots
  // Does it Crash?

// This article is a good resource for component testing
// https://www.sitepoint.com/test-react-components-jest/

// I do not get tests, and they do not get me.
// Leave testing to those with a knack for it. -KDC (2017)

const overrideObj = {
    fid: 0,
    date: '20170229'
}

// [describe] Optional - For logical grouping

// Can wrap Tests in either [test] or [it]

describe('FormDateContainer:', () => {

    test('Jest Snapshot', () => {
      const component = renderer.create(
        <FormDateContainer formParams={overrideObj} />
      )
      let tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDOM.render(<FormDateContainer formParams={overrideObj} />, div)
    })

    // Template below based on (a useful and helpful article):
        // https://medium.freecodecamp.com/the-right-way-to-test-react-components-548a4736ab22
        // https://medium.freecodecamp.com/@suchipi | Stephen Scott

        let props
        let mountedFormDateContainer
        const formDateContainer = () => {
            if (!mountedFormDateContainer) {
                mountedFormDateContainer = mount(
                    <FormDateContainer {...props} />
                    )
            }
            return mountedFormDateContainer
        }

        beforeEach(() => {
            props = {
                formParams: {
                    id: 0,
                    date: '20131021'
                }
            }
            mountedFormDateContainer = undefined
        })

    // All tests will go here

    it("always renders a div", () => {
        const divs = formDateContainer().find("div");
        expect(divs.length).toBeGreaterThan(0);
    })

    it("always renders a `FormDate`", () => {
        expect(formDateContainer().find(FormDate).length).toBe(1)
    })

    describe("when `showNativeDate` is true", () => {
        beforeEach(() => {
            props.showNativeDate = jest.fn()
        })

        it("[Native: True === Native is shown AND 'non-native' is not required]", () => {

            const datePicker = formDateContainer().find(FormDate),
                  isNativeProp = datePicker.props().showNativeDate,
                  isNativeDOM = !datePicker.find(".nativeDatePicker").get(0).style.display.includes('none'),
                  isNonRequired = datePicker.find(".fallbackDatePicker").find("#field-inputYear-0").get(0).hasAttribute('required')

            expect(isNativeProp).toEqual(isNativeDOM)       // Native: True === Native is not display:none (it is shown)
            expect(isNativeProp).not.toEqual(isNonRequired) // Native !== non-native is required
        })

        it("['Native' && field-date-0 === YYYYMMDD]", () => {
            const datePicker = formDateContainer().find(FormDate),
                  isNativeProp = datePicker.props().showNativeDate,
                  getValue = datePicker.find(".nativeDatePicker").find("#field-date-0").get(0).value,
                  getProp1 = datePicker.props().date,               // yyyymmdd
                  getProp2 = datePicker.props().fullDateYearLeft    // yyyy-mm-dd (value used in FormDate)

            if (isNativeProp) {
                expect(getValue).toBe(getProp2)
            }
        })
    })

    describe("when `showNativeDate` is false", () => {
        beforeEach(() => {
            props.showNativeDate = jest.fn()
        })

        it("[Non-native: True === Fallback is shown]", () => {

            const datePicker = formDateContainer().find(FormDate),
                  isNativeProp = datePicker.props().showNativeDate,
                  isNativeDOM = !datePicker.find(".nativeDatePicker").get(0).style.display.includes('none'),
                  isNonRequired = datePicker.find(".fallbackDatePicker").find("#field-inputYear-0").get(0).hasAttribute('required'),
                  isNonNative = !datePicker.find(".fallbackDatePicker").get(0).style.display.includes('none')

            if (isNonNative) {
                expect(!isNativeProp).toEqual(isNonNative)      // Non-native: True === Fallback div is not display:none (it is shown)
            }

            // wrapper.find('input').simulate('keyPress', {key: 'Enter'})
            // input.simulate('change', { target: { value: 'Hello' } })
            // input.simulate('blur');
        })

        it("[Not 'native' === 'non-native' is required]", () => {

            const datePicker = formDateContainer().find(FormDate),
                  isNativeProp = datePicker.props().showNativeDate,
                  isNativeDOM = !datePicker.find(".nativeDatePicker").get(0).style.display.includes('none'),
                  isNonRequired = datePicker.find(".fallbackDatePicker").find("#field-inputYear-0").get(0).hasAttribute('required'),
                  isNonNative = !datePicker.find(".fallbackDatePicker").get(0).style.display.includes('none')

            if (isNonNative) {
                expect(!isNativeProp).toEqual(isNonRequired)    // Not native === non-native is required
            }
        })

        // it("[Not 'native' && field-inputYear-0 === YYYY]", () => {
        //     const datePicker = formDateContainer().find(FormDate),
        //           isNativeProp = datePicker.props().showNativeDate
        //           // getValue = datePicker.find(".fallbackDatePicker").find("#field-inputYear-0").get(0).value,
        //           // getProp = datePicker.props().fullDateYearLeft
        //     // expect(!isNativeProp).toBe(true)
        // })

        // it("[Not 'native' && field-inputMonth-0 === MM]", () => {

        // })

        // it("[Not 'native' && field-inputDay-0 === DD]", () => {

        // })

    })
})