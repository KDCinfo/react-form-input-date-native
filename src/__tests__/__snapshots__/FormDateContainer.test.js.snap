// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDateContainer: Jest Snapshot 1`] = `
<form
  className="form-inline"
  data-id="form-0"
  onSubmit={[Function]}
>
  <div
    className="show-grid row"
  >
    <div
      className="field-col col-xs-12"
    >
      <div>
        <div
          className="nativeDatePicker"
          style={
            Object {
              "display": "block",
            }
          }
        >
          <div
            className="form-group has-feedback"
          >
            <label
              className="field-label"
              htmlFor="field-date-0"
            >
              <span>
                Date:
              </span>
            </label>
            <input
              className="form-control"
              disabled={false}
              id="field-date-0"
              name="inputDate"
              onChange={[Function]}
              placeholder="Date"
              required={true}
              type="date"
              value="2017-02-29"
            />
          </div>
        </div>
        <p
          className="fallbackLabel hide"
          style={
            Object {
              "display": "none",
            }
          }
        >
          Enter Date:
        </p>
        <div
          className="fallbackDatePicker"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <div
            className="form-group"
          >
            <label
              className="control-label"
              htmlFor="field-inputYear-0"
            >
              <span>
                Year
              </span>
            </label>
            <select
              className="form-control"
              disabled={false}
              id="field-inputYear-0"
              name="inputYear"
              onChange={[Function]}
              placeholder="Year"
              required={false}
              type={undefined}
              value={2017}
            >
              <option>
                2000
              </option>
              <option>
                2001
              </option>
              <option>
                2002
              </option>
              <option>
                2003
              </option>
              <option>
                2004
              </option>
              <option>
                2005
              </option>
              <option>
                2006
              </option>
              <option>
                2007
              </option>
              <option>
                2008
              </option>
              <option>
                2009
              </option>
              <option>
                2010
              </option>
              <option>
                2011
              </option>
              <option>
                2012
              </option>
              <option>
                2013
              </option>
              <option>
                2014
              </option>
              <option>
                2015
              </option>
              <option>
                2016
              </option>
              <option>
                2017
              </option>
              <option>
                2018
              </option>
              <option>
                2019
              </option>
              <option>
                2020
              </option>
              <option>
                2021
              </option>
              <option>
                2022
              </option>
              <option>
                2023
              </option>
              <option>
                2024
              </option>
              <option>
                2025
              </option>
              <option>
                2026
              </option>
              <option>
                2027
              </option>
            </select>
          </div>
          <div
            className="form-group"
          >
            <label
              className="control-label"
              htmlFor="field-inputMonth-0"
            >
              <span>
                Month
              </span>
            </label>
            <select
              className="form-control"
              disabled={false}
              id="field-inputMonth-0"
              name="inputMonth"
              onChange={[Function]}
              placeholder="Year"
              required={false}
              type={undefined}
              value="Feb"
            >
              <option>
                Jan
              </option>
              <option>
                Feb
              </option>
              <option>
                Mar
              </option>
              <option>
                Apr
              </option>
              <option>
                May
              </option>
              <option>
                Jun
              </option>
              <option>
                Jul
              </option>
              <option>
                Aug
              </option>
              <option>
                Sep
              </option>
              <option>
                Oct
              </option>
              <option>
                Nov
              </option>
              <option>
                Dec
              </option>
            </select>
          </div>
          <div
            className="form-group"
          >
            <label
              className="control-label"
              htmlFor="field-inputDay-0"
            >
              <span>
                Day
              </span>
            </label>
            <select
              className="form-control"
              disabled={false}
              id="field-inputDay-0"
              name="inputDay"
              onChange={[Function]}
              placeholder="Year"
              required={false}
              type={undefined}
              value={29}
            >
              <option>
                1
              </option>
              <option>
                2
              </option>
              <option>
                3
              </option>
              <option>
                4
              </option>
              <option>
                5
              </option>
              <option>
                6
              </option>
              <option>
                7
              </option>
              <option>
                8
              </option>
              <option>
                9
              </option>
              <option>
                10
              </option>
              <option>
                11
              </option>
              <option>
                12
              </option>
              <option>
                13
              </option>
              <option>
                14
              </option>
              <option>
                15
              </option>
              <option>
                16
              </option>
              <option>
                17
              </option>
              <option>
                18
              </option>
              <option>
                19
              </option>
              <option>
                20
              </option>
              <option>
                21
              </option>
              <option>
                22
              </option>
              <option>
                23
              </option>
              <option>
                24
              </option>
              <option>
                25
              </option>
              <option>
                26
              </option>
              <option>
                27
              </option>
              <option>
                28
              </option>
            </select>
          </div>
        </div>
      </div>
      <div
        className="entry-buttons"
      >
        <button
          className="btn btn-primary btn btn-default"
          disabled={false}
          type="submit"
        >
          Submit Form
        </button>
        <button
          className="btn btn-default btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Reset Form
        </button>
      </div>
    </div>
  </div>
</form>
`;
